lib/ansible/plugins/connection/qubes.py

```diff
@@ -0,0 +1,166 @@
+            cmd = cmd + "\n"
+        local_cmd = []
+
+        # For dom0
+        local_cmd.extend(["qvm-run", "--pass-io", "--service"])
+        if self.user != "user":
+            # Means we have a remote_user value
+            local_cmd.extend(["-u", self.user])
+
+        local_cmd.append(self._remote_vmname)
+
+        local_cmd.append(shell)
+
+        local_cmd = [to_bytes(i, errors='surrogate_or_strict') for i in local_cmd]
+
+        display.vvvv("Local cmd: ", local_cmd)
+
+        display.vvv("RUN %s" % (local_cmd,), host=self._remote_vmname)
+        p = subprocess.Popen(local_cmd, shell=False, stdin=subprocess.PIPE,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        # Here we are writing the actual command to the remote bash
+        p.stdin.write(to_bytes("%s" % cmd, errors='surrogate_or_strict'))
+        stdout, stderr = p.communicate(input=in_data)
+        stdout = to_bytes(stdout, errors='surrogate_or_strict')
+        stderr = to_bytes(stderr, errors='surrogate_or_strict')
+        return p.returncode, stdout, stderr
+
+    def _connect(self):
+        """No persistent connection is being maintained."""
+        super(Connection, self)._connect()
+        self._connected = True
+
+    @ensure_connect
+    def exec_command(self, cmd, in_data=None, sudoable=False):
+        """Run specified command in a running QubesVM """
+        super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
+
+        display.vvvv("CMD IS: %s" % cmd)
+
+        rc, stdout, stderr = self._qubes(cmd)
+
+        display.vvvvv("STDOUT %r STDERR %r" % (stderr, stderr))
+        return rc, stdout, stderr
+
+    def put_file(self, in_path, out_path):
+        """ Place a local file located in 'in_path' inside VM at 'out_path' """
+        super(Connection, self).put_file(in_path, out_path)
+        display.vvv("PUT %s TO %s" % (in_path, out_path), host=self._remote_vmname)
+
+        with open(in_path, "rb") as fobj:
+            source_data = fobj.read()
+
+        retcode, dummy, dummy = self._qubes('cat > "{0}"\n'.format(out_path), source_data, "qubes.VMRootShell")
+        # if qubes.VMRootShell service not supported, fallback to qubes.VMShell and
+        # hope it will have appropriate permissions
+        if retcode == 127:
+            retcode, dummy, dummy = self._qubes('cat > "{0}"\n'.format(out_path), source_data)
+
+        if retcode != 0:
+            raise RuntimeError('Failed to write target file {0}'.format(out_path))
+
+    def fetch_file(self, in_path, out_path):
+        """Obtain file specified via 'in_path' from the container and place it at 'out_path' """
+        super(Connection, self).fetch_file(in_path, out_path)
+        display.vvv("FETCH %s TO %s" % (in_path, out_path), host=self._remote_vmname)
+
+        # We are running in dom0
+        cmd_args_list = ["qvm-run", "--pass-io", self._remote_vmname, "'cat {0}'".format(in_path)]
```

Comment:
Does qvm-run resolve the quotes around "cat in_path" itself?  If it just needs to be a single argument, you should be able to remove the quotes and it will still work (because you're now passing a list to subprocess.Popen).

Reply:
qvm-run just require this to be a single argument, no extra quotes are needed

Reply:
No, it does not work that way. https://www.qubes-os.org/doc/copy-from-dom0/#copying-to-dom0 has the details on this.

Reply:
<nod> So this should work:
``` python
         cmd_args_list = ["qvm-run", "--pass-io", self._remote_vmname, "cat {0}".format(in_path)] 
```

(The qubes doc example is being parsed by a shell so quotes are needed to make sure the "cat filename" is a single argument.  We're passing a list of arguments directly to the platform's C library so we do not need the quotes.  We simply make sure that the "cat filename" is a single entry in the argument list.)